### ISSUE TITLE
handling clock skews in distributed systems

### ISSUE THREAD

**Issue by asldevi** [hypertrace/hypertrace#139]
## Use Case

Handle clock skews in traces at query service as different spans may come from different hosts in the distributed systems, which might be otherwise confusing to the end users. It'd be great if this can be toggled on/off from the UI.

## Proposal
Adjusting spans based on a config param like [Jaeger does]((https://www.jaegertracing.io/docs/1.21/deployment/#clock-skew-adjustment) would be very useful.

**Comment by jcchavezs** [comment]
+1

**Comment by suddendust** [comment]
I am picking this up.

**Comment by suddendust** [comment]
@rish691  Which component should handle this ideally? All attribute conversions from the Gateway service are being done in the GQL layer. Should this be handled there as well?

**Comment by rish691** [comment]
So any query layer change would generally go through gql and gateway & one of entity/query service (conditionally)

But I guess we should be starting with the design proposal and then get to which component do the changes belong to

**Comment by suddendust** [comment]
## Proposal

Legend:

| Symbol      | Meaning |
| ----------- | ----------- |
| _d<sub>parent<sub>_      | Parent span duration       |
| _d<sub>child<sub>_   | Child span duration        |
|_S<sub>parent<sub>_| Start time of parent span|
|_E<sub>parent<sub>_| End time of parent span |
|_S<sub>child<sub>_| Start time of child span|
|_E<sub>child<sub>_| End time of child span|

The first step is to determine whether a parent-child span pair is eligible for adjustment. We don't do any adjustment in the following cases:

1. If a child span already fits in its parent span. That is, if _S<sub>child<sub>_  >= _S<sub>parent<sub>_ && _E<sub>child<sub>_ <= _E<sub>parent<sub>_. It follows that  _d<sub>parent<sub>_ >= _d<sub>child<sub>_ in this case.
2.  If _d<sub>child<sub>_  > _d<sub>parent<sub>_, and _S<sub>child<sub>_ >= _S<sub>parent<sub>_. The following diagram illustrates this case:

![Untitled Diagram](https://user-images.githubusercontent.com/84911643/123052328-554bb900-d420-11eb-8ec0-c6f6227acdff.png)

The above scenario can happen in case of async _fire-and-forget_ calls, or if the client timed out before it got a response from the server. 

We adjust in the following cases:

1. If _d<sub>child<sub>_  > _d<sub>parent<sub>_, and _S<sub>child<sub>_ < _S<sub>parent<sub>_. In this case, a child starts before its parent as:

![child_before_parent](https://user-images.githubusercontent.com/84911643/123056720-cdb47900-d424-11eb-8382-02bbb82a7b61.png)

In this case, we have to simply move the child span right by _S<sub>parent<sub>_ - _S<sub>child<sub>_.

2. If _d<sub>child<sub>_  < _d<sub>parent<sub>_, and if either of the two conditions hold:  (_S<sub>child<sub>_ > _S<sub>parent<sub>_ && _E<sub>child<sub>_ > _E<sub>parent<sub>_), or (_S<sub>child<sub>_ < _S<sub>parent<sub>_ && _E<sub>child<sub>_ < _E<sub>parent<sub>_).

Condition1:

![case3](https://user-images.githubusercontent.com/84911643/123057690-b629c000-d425-11eb-93f7-d40cc8fba634.png)

Condition2:

![case4](https://user-images.githubusercontent.com/84911643/123057917-e709f500-d425-11eb-9f66-4273088db3b2.png)

In both the case, we are aiming for something like:

![case5](https://user-images.githubusercontent.com/84911643/123058651-a9f23280-d426-11eb-9180-818646951167.png)

_d_ can be calculate using the following heuristic:


_d_ = (_d<sub>parent<sub>_ - _d<sub>child<sub>_) / 2.

Ultimately, we have:

_S<sub>parent<sub>_ + _d_ = _S<sub>child<sub>_ + _adjustment_. Once _adjustment_ is known, we can adjust the skew as:

_S<sub>child<sub>_ := _S<sub>child<sub>_ + _adjustment_. For condition1, _adjustment_ will be negative, and for condition2, it will be positive.

This heuristic is basically what Jaeger uses.

@rish691  please review.

**Comment by kotharironak** [comment]
Isn't the `condition 1` of ` d(child) < d(parent), and if either of the two conditions hold: (Schild > Sparent && Echild > Eparent), or (Schild < Sparent && Echild < Eparent)` possible in async case too? 
 
Referring to below image:
Conditon 1 - ` Schild > Sparent && Echild > Eparent`
![condition1](https://user-images.githubusercontent.com/53209990/123099802-24837800-d450-11eb-8304-d4e83abe87df.png)

**Comment by suddendust** [comment]
Yes. Looks like there is no way to find if condition 1 is a _follows-from_  or a genuine case of skew. I don't think this should be adjusted.

**Comment by suddendust** [comment]
The heuristic aside, what would be the better way to implement this:

1. Do all of these adjustments in the query layer. Having a sort of span transformation pipeline in which you can register transformers. This approach is extensible in the sense that if tomorrow if we want to modify spans in a totally different story, you just need to register one more handler. However, doing this in the query layer can have some overhead, especially if the number of spans is very large.

2. Doing this while ingesting and saving adjusted timestamps in the DB. IMO, since this is an optional feature, storing these values in the DB (or altering start_time and end_time) would be doing too much for too little (we will also need to alter the table structure and it wouldn't work for existing spans). I think any span transformations should occur in the query layer (unless the transformation pipeline takes too much time).

@kotharironak @rish691 please let me know you thoughts.

**Comment by jcchavezs** [comment]
> Do all of these adjustments in the query layer. Having a sort of span transformation pipeline in which you can register transformers. This approach is extensible in the sense that if tomorrow if we want to modify spans in a totally different story, you just need to register one more handler. However, doing this in the query layer can have some overhead, especially if the number of spans is very large.

I am inclined to say this should happen in query layer, we already do a bunch of modifications in user data and doing this in storage can mess up user data with no return unless we do in a sort of view model for this and not in the source of truth.

Regarding this proposal I think we should make it clear all the adjustments should happen in the outbound edge, meaning that **local spans** shouldn't be adjusted (this is, spans that share service with the parent). 

Another comment on this is how about the adjustment of two siblings spans where one is in a host with clockskew and the other does not, how do we resolve the adjustment?

**Comment by suddendust** [comment]
@jcchavezs thank you for reviewing the proposal. Regarding the implementation, I think the better place to do this is in gateway-service once it fetches the spans (SpanService.java to be more specific). Once they are fetched, whatever transformations can be applied on spans. What do you think?

I actually didn't get what you mean by "spans that share service with the parent". Could you please explain this scenario?

**Comment by jcchavezs** [comment]
A span can have several childs, isn't? e.g. imagine an endpoint that does two calls to a database to prepare the response, both calls have the same parent isn't? All the calculations I see in the proposal only consider one child per parent which in practice isn't the only case. Does it make sense?

**Comment by suddendust** [comment]
Yes, it makes sense, thank you for the explanation. I am not really sure relationships among siblings can be maintained while adjusting the spans. Jaegar doesn't seem to handle this as well. See https://github.com/jaegertracing/jaeger/issues/961. See the following image from the same issue:

<img width="885" alt="Screenshot 2021-07-01 at 8 30 19 PM" src="https://user-images.githubusercontent.com/84911643/124146382-55485a80-daab-11eb-8fcb-41127098b023.png">

Notice the two siblings. They are sequential (one after the other). However, this information is totally lost after the adjustment. Jaeger just warns users about the adjustment and gives them an option on the UI to turn it off. Please let me know how you think this can be handled better.

**Comment by kotharironak** [comment]
@jcchavezs @suddendust I am sorry for the late reply. If the strategy only applies to a specific API (say span) at query time, that can lead to confusion in terms of metrics. e.g, let's consider that one of the services (X) is having a clock skew issue. So, the latency metric for this service (X)  will be calculated using the original startTime, endTime of its entry span. And so, it may not match with overall trace latency or service latency but its individual spans will show different duration based on the applied query time clock skew strategy. And so, I was thinking of doing this at ingestion time by readjusting startTime/endTime based on strategy (either in raw-spans-grouper / enricher). If required, we can store original data into different columns too.

**Comment by suddendust** [comment]
@kotharironak the child spans are simply shifted temporally (so basically you apply the same adjustment to the start time and the end time) wrt their parents. The duration doesn't change. Will this still cause a discrepancy?

**Comment by kotharironak** [comment]
> @kotharironak the child spans are simply shifted temporally (so basically you apply the same adjustment to the start time and the end time) wrt their parents. The duration doesn't change. Will this still cause a discrepancy?

Yah, if they are shifted, it will not impact metrics - latency, call rate, error rate

**Comment by rish691** [comment]
> d can be calculated using the following heuristic:
d = (dparent - dchild) / 2.
Ultimately, we have:
Sparent + d = Schild + adjustment. Once adjustment is known, we can adjust the skew as:
Schild := Schild + adjustment. For condition1, adjustment will be negative, and for condition2, it will be positive.

Whats the intuition behind setting d as `(dparent - dchild) / 2` ? 

I see this adjustment causing lots of confusion (in the issue that you point and few others)

Why not say do something like linearly fit the all the child spans in the parent boundary? 

The more important question before discussing on the correctness of the heuristics is around `what is the purpose of these clock-skew adjustment`?
From what I could gather (from jaeger docs and issue), is that it helps in transforming the trace waterfall view to appear logical. 
If that is the case then the way the current heuristics handle the case for multiple child spans doesn't seem to solving the problem.

**Comment by jcchavezs** [comment]
Zipkin does all this in the UI, I think that solution is less problematic
because we don't need to touch anything in processing logic. See
https://github.com/openzipkin/zipkin/blob/4a690eb1486eea8d5eaf10e47f1ea350b3ad0959/zipkin-lens/src/zipkin/clock-skew.js
and see the tests
https://github.com/openzipkin/zipkin/blob/4a690eb1486eea8d5eaf10e47f1ea350b3ad0959/zipkin-lens/src/zipkin/clock-skew.test.js#L75

Can we do the same?

**Comment by JBAhire** [comment]
First, I will answer @rish691 's question

> The more important question before discussing the correctness of the heuristics is around what is the purpose of these clock-skew adjustments?

Handling clock skew is important when it comes to specifically frontend telemetry. Frontend telemetry data can be skewed because of the quality of the platforms. For example, Mobile devices and desktops can have inaccurate clocks;  so, the time reported by these devices can be skewed by seconds or minutes compared to time reported by the backend. This can cause situations like @suddendust mentioned above like parent span starts after child span and others.

Use-case wise there are two important use-cases I see and those are,
- More accurate frontend telemetry data and waterfall view (this is contextual and relative but talking at high level.)
- Finding network performance bottlenecks (and this one is very important in critical services and operations. )

Now, I might be repeating some pointers we discussed but to summarise, there are several ways to solve this problem at a high level.
1. Tracing solutions can measure this clock skew using adaptations of the Network Time Protocol. Here, we need to record timestamps on both tracing libraries and within tracing  implementation and comparing both values can help us in estimating skew. Then we just simply add or subtract this skew from the timestamp of the frontend span. This is similar to what @suddendust is talking about and this is effective in most of the cases.

      The main problem with this one is unpredictability which @jcchavezs also brought into question. So whenever it fails due to some uncertainty on device end, the results can be very confusing and less accurate than original timestamps. We can solve it by providing a way for developers to switch between original and adjusted timestamps. That is how Jaeger also solves it, I guess. (They don't actually solve this in UI as of now: https://github.com/jaegertracing/jaeger-ui/issues/197 [Thanks @suddendust for pointing to the issue.])

2. A second method is to use casual relationships. So, there are some hard facts when it comes to distributed tracing and casual relationships and we can leverage that information to find and adjust clock skews between frontend and backend services. One such fact is a client  span  must  start  before  its  associated  server  span  starts.  This might not always be the case but in many apps server spans end before associated client spans and we can use these relationships to adjust the timestamps to maintain these factual invariable relationships.

      Although this method is more predictable, this is too coarse. It won't give you actual skew or variation but just provide you the upper bound. We can't really calculate important information like network time consumption using it.

3. Third way ideally will be using a combination of the above two methods to mitigate confusion which the first one might bring and to increase accuracy which the second one lacks. (@jcchavezs can give a better explanation but this is kinda similar to what he was pointing towards with Zipkin, I guess.)


Going back to **How we want to implement this with Hypertrace?**
- I agree with @suddendust that it makes more sense to add this logic in the Gateway service. To your point @jcchavezs , our UI doesn't handle any logic at this moment. It just pulls and shows the data so, ideally it will be better to add this in Gateway service and use it. (@kotharironak, @rish691 and I had small chat about this in the morning so folks please feel free to chip in. )
- The first/ default view in Hypertrace should be without clock skew adjustment and then we should provide a button in top left to see the adjusted view. Ideally, I will prefer it if we don't show it at all in API Trace view and only let people adjust for skews in full trace views because it might cause confusion in API trace view.

I tried to put together high-level design below (open for discussion of course):


| <img width="1308" alt="Screenshot 2021-07-12 at 2 03 11 PM" src="https://user-images.githubusercontent.com/26570044/125256261-ec1dde00-e319-11eb-97da-33695826a2af.png"> |
|:--:|
| *The clock skew button beside expand and collapse button can help you switch between adjusted and unadjusted views* |
